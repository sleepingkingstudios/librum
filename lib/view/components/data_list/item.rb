# frozen_string_literal: true

module View::Components
  # Component for rendering an item in a data list.
  class DataList::Item < View::Components::DataField
    # @param data [Hash{String=>Object}] the data used to render the list item.
    # @param field [View::Components::DataField::FieldDefinition] the
    #   configuration object for rendering the list item.
    def initialize(data:, field:, **options)
      super(
        data:  data,
        field: field,
        **options
      )
    end

    private

    def render_envelope?
      if field_value.blank? && default.present?
        return !default_value.is_a?(ViewComponent::Base)
      end

      !field_value.is_a?(ViewComponent::Base)
    end
  end
end
