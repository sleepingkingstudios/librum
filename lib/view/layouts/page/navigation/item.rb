# frozen_string_literal: true

module View::Layouts
  # Component for rendering a page navigation item.
  class Page::Navigation::Item < ViewComponent::Base
    extend Forwardable

    # @param item [View::Layouts::Page::Navigation::ItemDefinition] the
    #   navigation item to render.
    def initialize(item:)
      super()

      @item = item
    end

    def_delegators :@item,
      :items,
      :label,
      :url

    # @return [View::Layouts::Page::Navigation::ItemDefinition] the navigation
    #   item to render.
    attr_reader :item

    private

    def dropdown?
      items.present?
    end
  end
end
