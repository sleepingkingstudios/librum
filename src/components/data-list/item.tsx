import * as React from 'react';
import {
  capitalize,
  snakeCase,
} from 'lodash';

import { isElement } from '@utils/react-utils';
import type { Literal } from '@utils/types';

type DataListItemProps = {
  defaultValue?: JSX.Element | string;
  label: string;
  value: JSX.Element | Literal;
};

const formatLabel = (label: string): string => {
  return snakeCase(label).split('_').map(capitalize).join(' ');
};

const formatValue = (value: Literal): string => {
  return value.toString();
};

const renderValue = ({
  defaultValue,
  value,
}: {
  defaultValue: JSX.Element | string | null,
  value: JSX.Element | Literal,
}): JSX.Element | string => {
  if (isElement(value)) {
    const element = value as JSX.Element;

    return element;
  }

  if (value !== null) {
    const literal = value as Literal;

    return formatValue(literal);
  }

  return defaultValue;
};

export const DataListItem = ({
  defaultValue = null,
  label,
  value,
}: DataListItemProps): JSX.Element => (
  <div className="data-list-item">
    <dt className="data-list-label">
      { formatLabel(label) }
    </dt>
    <dd className="data-list-value">
      { renderValue({ defaultValue, value }) }
    </dd>
  </div>
);
