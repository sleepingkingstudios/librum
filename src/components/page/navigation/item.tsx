import * as React from 'react';
import {
  Link,
  useLocation,
} from 'react-router-dom';

export interface NavigationItemProps {
  label: string;
  url: string;
}

export const NavigationItem = ({ label, url }: NavigationItemProps): JSX.Element => {
  const { pathname } = useLocation();

  if (pathname === url) {
    return (
      <li className="navigation-item">
        <p className="navigation-item-link-disabled text-muted">{ label }</p>
      </li>
    );
  } else {
    return (
      <li className="navigation-item">
        <Link className="navigation-item-link" to={url}>{ label }</Link>
      </li>
    );
  }
};
