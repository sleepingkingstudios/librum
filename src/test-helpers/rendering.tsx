import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
