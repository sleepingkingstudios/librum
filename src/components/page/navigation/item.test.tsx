import * as React from 'react';

import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';

import { NavigationItem } from './item';
import { render } from '@test-helpers/rendering';

describe('<NavigationItem>', () => {
  const defaultProps = {
    label: 'Go Back',
    url: '/back',
  };

  describe('when the pathname does not match the url', () => {
    it('should render the link', async () => {
      const { getByRole } =
        render(<NavigationItem {...defaultProps} />, { router: true });

      const link = getByRole('link', { name: defaultProps.label });

      expect(link).toBeVisible();

      await userEvent.click(link);

      expect(link).not.toBeVisible();
    });

    it('should match the snapshot', () => {
      const { asFragment } =
        render(<NavigationItem {...defaultProps} />, { router: true });

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when the pathname matches the url', () => {
    it('should not render a link', () => {
      const {
        getByText,
        queryByRole,
      } = render(
        <NavigationItem {...defaultProps} />,
        {
          initialEntries: [defaultProps.url],
          router: true,
        },
      );

      expect(queryByRole('link')).toBeNull();

      expect(getByText(defaultProps.label)).toBeVisible();
    });

    it('should match the snapshot', () => {
      const { asFragment } =
        render(
          <NavigationItem {...defaultProps} />,
          {
            initialEntries: [defaultProps.url],
            router: true,
          },
        );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
