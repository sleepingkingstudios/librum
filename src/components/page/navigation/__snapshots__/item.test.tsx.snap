// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavigationItem> when the pathname does not match the url should match the snapshot 1`] = `
<DocumentFragment>
  <li
    class="navigation-item"
  >
    <a
      class="navigation-item-link"
      href="/back"
    >
      Go Back
    </a>
  </li>
</DocumentFragment>
`;

exports[`<NavigationItem> when the pathname matches the url should match the snapshot 1`] = `
<DocumentFragment>
  <li
    class="navigation-item"
  >
    <p
      class="navigation-item-link-disabled text-muted"
    >
      Go Back
    </p>
  </li>
</DocumentFragment>
`;
